### ISSUE TITLE
System settings as sufficient techniques?

### ISSUE THREAD

**Issue by julianmka** [w3c/matf#67]
We've had a few conversations about whether native mobile operating system settings such as high contrast, extended time for toasts, reduced motion/animation, and others might satisfy related WCAG success criteria. 

<details><summary>Conversation between @alastc and @JJdeGroot in #32. There is precedent in WCAG for accepting system-level settings as sufficient.</summary>
<p>
JJ:
For 2.2.1 Timing Adjustable, the user is able to adjust some timing related settings at the system level on Android and iOS

Is this an acceptable "mechanism"? Note 1 in the definition mentions: "The mechanism may be explicitly provided in the content, or may be relied upon to be provided by either the platform or by user agents, including assistive technologies."
This not only includes user agents, but also the platform, in this case Android and iOS; therefore it's a sufficient technique to conform?
But... users might not be aware of such system settings, because they can be buried multiple levels deep

AC:
I’m a bit confused, do you mean for 2.2.2 Pause, Stop, Hide? 2.2.1 doesn’t mention a mechanism. If there were a system level thing that allowed users to turn-off/adjust/extend, I think that could be used.

JJ:
Sorry, the note is indeed for 2.2.2. There is a system setting on new Android versions to increase the time limit of [toasts](https://developer.android.com/guide/topics/ui/notifiers/toasts).

AC:
In which case, I think you could allow for a system level thing that allowed users to turn-off/adjust/extend.

JJ: Okay, the question is when is a setting considered widely-available? Would auditors need to write this down in the "Accessibility support baseline" section? E.g. some settings are only available on newer Android version, and some settings be buried deep in the settings screens.

AC:
In WCAG 2.x we’ve taken that on a case by case basis but, for example, a system setting is sufficient to pass Animation from interactions in [C39](https://www.w3.org/WAI/WCAG22/Techniques/css/C39). This sounds similar.

This topic is a constant discussion, there isn’t a clear and concrete answer for a world-wide standard. However, if the setting is free (once the base device and OS are paid for), and on more than 66% of used devices it should be a fairly safe option.
</p>
</details>

**Comment by AlainVagner** [comment]
For me this is a little bit of a chicken and egg issue. On one hand, the users may not be aware that some settings exist, on the other hand, if systems settings are never considered as a sufficient technique, they will be rarely supported and thus leading to a limited usage. At some point, we should consider if it makes sense to promote these settings, in order to avoid having different and custom settings in every app.

Not sure how we will evaluate if a setting is widely available though. 
Does the setting need to be widely available on iOS and Android at the same time to be considered as sufficient?

I have listed here the settings we consider when testing different SC on mobile apps in Luxembourg:

- 1.4.1
    - Differentiate Without Colour on iOS 
- 1.4.3 
    - Increase Contrast on iOS 
- 1.4.4 
    - Larger Text on iOS
    - Font size on Android
- 2.4.7
    - Focus styles on iOS

**Comment by JJdeGroot** [comment]
Insight I had during a discussion:

What if we check who has to do the work when enabling system settings.

E.g. 
- When users activates the zoom function, they have to do the work to zoom text
- When a user activates the screen reader, they have to use gestures, but developers also have to implement accessibility support
- When users activate high contrast, the developers have to take the settings into account, otherwise the text contrast remains unchanged
- When users activate dark mode, the developers have to implement a dark theme, otherwise the app remains in light colors
- When users activate text scaling, developers have to implement larger text scaling, otherwise text remains normal sized.

When categorizing assistive technology/feature when can check if activating burdens the user or the app developer.

If the developers have to do the work, using the system setting is sufficient to pass.

**Comment by julianmka** [comment]
I was thinking about a similar rule but kept getting stuck on handling something like increased contrast settings. I like this idea of establishing a practice of listing out supported accessibility settings, @JJdeGroot. 

And for more controversial settings like increased contrast, we could still include a note that meeting the SC without the user having to enable an accessibility feature is preferable. This feels like a good way forward.

**Comment by PaulvanWorkum** [comment]
@jeroendevrind, that is true, but so far I understand this only works on unadjusted keyboard focus indicators. If a developer did not change the standard focus indicator the indicator already falls under the exception.

**Comment by detlevhfischer** [comment]
> Does the setting need to be widely available on iOS and Android at the same time to be considered as sufficient?

To us this is a non-issue since we would audit iOS and Android apps independently,

**Comment by detlevhfischer** [comment]
@JJdeGroot wrote:
> If the user has to do the work, the system setting does not pass. E.g. using zoom function to enlarge text is a fail.

For some things like zoom / magnification functions, I do not agree. As far as I am aware, many companies providing app audits today consider 1.4.4 a PASS by default because of that OS AT function (probably recommending the support of OS settings based text sizing at the same time).  The normative text 1.4.4 does not differentiate between body text and other types of text like icon labels, headings etc., so in many cases making **all** text scale to 200% would break layouts / have undesirable consequences - so content would nominally fail 1.4.4 if not everything is equally resizable.

Listing Accessibility and compatibility features is good but I don't quite see the point of listing zoom / magnification since it always works, independent of the app. So wouldn't people relying on it be aware if the function anyway?

**Comment by detlevhfischer** [comment]
@PaulvanWorkum 

I accidentally editied your comment to @jeroendevrind above and have restored it but I am not sure there wasn't more you wrote. Sorry about that! (Why can I even edit other peoples comments??)

@PaulvanWorkum  wrote:
> So far I understand this only works on unadjusted keyboard focus indicators. If a developer did not change the standard focus indicator the indicator already falls under the exception.

Are you referring here to the exception in 1.4.11: "except for inactive components or where the appearance of the component is determined by the user agent and not modified by the author"?

I have so far understood "state" to be referring to attributes proper of the component, like checked or selected, but it is true that the definition includes focus. So all the light grey and light blue keyboard focus indications in apps would automatically pass via this exception in 1.4.11?

**Comment by PaulvanWorkum** [comment]
@detlevhfischer , I also didn't know it was possible to adjust other peoples comments. Not a problem.

Definition of state according to WCAG:
> state
> dynamic property expressing characteristics of a user interface component that may change in response to user action or automated processes
> 
> States do not affect the nature of the component, but represent data associated with the component or user interaction possibilities. Examples include **focus**, hover, select, press, check, visited/unvisited, and expand/collapse.

The developer cannot adjust the appearance of the focus-state of the keyboard at the moment. Although the developer can adjust the appearance of the component in other ways.

At Abra we indeed do not fail our clients at this at the moment, but we give it as a suggestion for improvement. We do not fail on it because the author did not adjust the appearance of the focus-state. It is determined by the "user agent".

**Comment by detlevhfischer** [comment]
@PaulvanWorkum 

> The developer cannot adjust the appearance of the focus-state of the keyboard at the moment

Hmm... I am not sure I understand. On [appt.org](https://appt.org/en/docs/ios/samples/accessibility-focus-indicator) you state for iOS:
> On iOS, you can adjust colors when an element receives focus. 

And there is the same sentence in the Android section.

**Comment by PaulvanWorkum** [comment]
> @JJdeGroot wrote:
> 
> > If the user has to do the work, the system setting does not pass. E.g. using zoom function to enlarge text is a fail.
> 
> For some things like zoom / magnification functions, I do not agree. As far as I am aware, many companies providing app audits today consider 1.4.4 a PASS by default because of that OS AT function (probably recommending the support of OS settings based text sizing at the same time). (Having said that, of course this practice could be challenged.)
> 
> The normative text 1.4.4 does not differentiate between body text and other types of text like icon labels, headings etc., so in many cases making **all** text scale to 200% would break layouts / have undesirable consequences - so content would nominally fail 1.4.4 if not everything is equally resizable.
> 
> Listing Accessibility and compatibility features is good but I don't quite see the point of listing zoom / magnification since it always works, independent of the app. So wouldn't people relying on it be aware if the function anyway?

WCAG 1.4.4 understanding doucment states:

> The intent of this Success Criterion is to ensure that visually rendered text, including controls and labels using text, can be made larger so that it can be read more easily by people with milder visual impairments, without requiring the use of assistive technology (such as a screen magnifier). 

> The working group feels that 200% is a reasonable accommodation that can support a wide range of designs and layouts, and complements older screen magnifiers that provide a minimum magnification of 200%. Above 200%, zoom (which resizes text, images, and layout regions and creates a larger canvas that may require both horizontal and vertical scrolling) may be more effective than text resizing. Assistive technology dedicated to zoom support would usually be used in such a situation, and may provide better accessibility than attempts by the author to support the user directly.

So zoom is not intended to be used to conform to 1.4.4.

WCAG 1.4.4 states:
> text can be resized without assistive technology up to 200 percent

I think Zoom is an assistive technology. Also in the definition of @JJdeGroot it remains Assistive technology as intended by 1.4.4.

Using zoom to pass 1.4.4 by default is to my opinion incorrect. From a guideline perspective, but also from a user perspective, as up to 25% of users is using this setting in the Netherlands.

In my opinion the need to use zoom to scale text to 200% is not sufficient to pass 1.4.4.

But I am very curious how this is done by others.

**Comment by detlevhfischer** [comment]
@PaulvanWorkum  The 1.4.4 normative text is agnostic regarding ways to achieve 200%. Zooming is one of them: Sufficient techniques include zoom, by for the most common way to meet 1.4.4 for web content.  The statment that above 200%, zoom may be more effective does not imply that up to 200%, zoom cannot be used to conform.

As to "is native Zoom AT?" question: Often (e.g. in the EN 301 549) the OS for software is regarded as equivalent to the UA for web content. So when Zoom as a built-in UA feature allows you to meet 1.4.4 in the web context, the implication is not far-fetched that zoom as a built-in OS function operates basically in an equivalent way and can meet 1.4.4 here. 

The alternative (requiring all text to be scalable to 200% via text size settings) will in practice mean that hardly any app will ever conform (because due to the layout-busting consequences of scaling up text, it is hardly ever applied across ALL text content).

**Comment by JJdeGroot** [comment]
@detlevhfischer 
> > The developer cannot adjust the appearance of the focus-state of the keyboard at the moment
> 
> Hmm... I am not sure I understand. On [appt.org](https://appt.org/en/docs/ios/samples/accessibility-focus-indicator) you state for iOS:
> 
> > On iOS, you can adjust colors when an element receives focus.
> 
> And there is the same sentence in the Android section. But I guess it is the difference between changing the color of the elements focused (e.g. changing background color of a button when focused) vs . being able to change the color of the system-provided focus mechanism (light grey or blue background, or focus ring in accent color).

To confirm: yes the different is the appearance of focused elements vs the appearance of the keyboard focus indicator. 

On web, you can use `:focus` to change the focus appearance, including the keyboard focus indicator.

 On Android/iOS you can use accessibility events to keep track of focus and change the focus appearance, but only of the element itself. Only users can change the appearance of the keyboard focus indicator (e.g. high contrast version)

And as Paul mentioned, because of the user-agent exception in 1.4.11 we do not fail the low contrast of the default keyboard focus indicator.

@detlevhfischer 
> The alternative (requiring all text to be scalable to 200% via text size settings) will in practice mean that hardly any app will ever conform (because due to the layout-busting consequences of scaling up text, it is hardly ever applied across ALL text content).

My experience is that apps can reasonably support 200% text scaling for main content. Text scaling is problematic in the navigation bar and tab bar. iOS has the Large Content Viewer to show the text enlarged in a pop-up (similar to zoom). Android doesn't have this feature at this moment. In those places, it's acceptable to need to use zoom.

In my opinion zoom should not be sufficient to pass 1.4.4 **if the platform does have support for text scaling.**

[W3C, Low Vision Task Force](https://w3c.github.io/low-vision-a11y-tf/requirements.html#text-size)
> **User Need - Text Size:**
> Users can change the text size (font size) of all text, without zooming the entire interface. 

[Google, Android UI Design Guide](https://developer.android.com/design/ui/mobile/guides/foundations/accessibility)
> Follow these guidelines to design for vision accessibility.
> To allow users to adjust the font size, specify font size in [scalable pixels (sp)](https://developer.android.com/training/multiscreen/screendensities#TaskUseDP)

[Apple, Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/typography)
> Support Dynamic Type. When you support Dynamic Type — a system feature that lets people choose the size of visible text on their device — your app or game can respond appropriately when people adjust text to a size that works for them. 

Allowing zoom to meet 1.4.4 does not align with industry best-practices and user needs.

**Comment by detlevhfischer** [comment]
> In my opinion zoom should not be sufficient to pass 1.4.4 if the platform does have support for text scaling.

@JJdeGroot Well, in terms of the normative language, and given my argument above which is confirmed by the WCAG2ICT take, OS zoom **is** sufficient to meet 1.4.4, even if supporting text scaling is certainly a recommended best practice.

**Comment by JJdeGroot** [comment]
@detlevhfischer Yes, I do agree that zoom is sufficient when you follow the language in WCAG2ICT. However, given that WCAG2ICT itself is not-normative, just like guidance of the MATF, I think that in a mobile-context, we should deviate from language in WCAG2ICT when needed. In my opinion, this would be one of those cases. 

Given the enormous amounts of accessibility features and assistive technologies that ship with Android and iOS, an app would automatically pass many Success Criteria; while burdening the user.

We need to strike a balance between what's reasonable to expect from developers and what's reasonable to expect from users.

That will be hard to define without cherry picking certain features such as Zoom.

**Comment by JJdeGroot** [comment]
FYI: I've closed https://github.com/w3c/matf/issues/64 because it's a duplicate of the system settings topic.

**Comment by detlevhfischer** [comment]
@JJdeGroot  wrote:
> However, given that WCAG2ICT itself is not-normative, just like guidance of the MATF, I think that in a mobile-context, we should deviate from language in WCAG2ICT when needed. In my opinion, this would be one of those cases.

Sure. It is just important to be clear that non-normative docs cannot extend normative requirements as worded. Currently, our normative point of departure for app audits is EN 301 549 V.3.2.1 clause 11 which has the same uambiguous text "...or that the application works with the platform features that meet this requirement." So there is no way we could normatively fail an app relying on OS zoom if our normative reference is the current EN. Of course an audit procedure may deviate from the EN and ask for more -- that can be useful but it is then not a EN 301 549 conformance audit.

BTW the next EN version draft V4.1.1b (so far) does not change that note in 11.1.4.4.1 Resize text (open functionality). This may change of course, and it may be the time to raise an issue in the [ETSI Github](https://labs.etsi.org/rep/HF/en301549/-/issues) to request such a change.

**Comment by JJdeGroot** [comment]
@detlevhfischer - Yes I understand the boundaries of our guidance. I did not realize that both updated WCAG2ICT and the draft of EN 301 549 contains this new clause to allow platform features for 1.4.4.

Android and iOS have two "Zoom" features: 
1) Screen zoom, which enlarges all content on the screen and this does result in reflow. Developers have to account for small screens, otherwise there will be text truncation and overlapping elements.
2) Screen magnify, which enlarges a section of the screen **without** reflow. Users have to scroll horizontally and vertically at the same time to read content.

In my opinion, it is not desirable that Magnification is sufficient, you will encounter the problems described in 1.4.10 Reflow, e.g. scrolling horizontally and vertically at the same time. This burdens the user.

I have fewer problems with screen zooming, but I'm not sure if it achieves 200% text scaling. This burdens the developers, but mainly to support smaller screens, not for supporting larger text; because the operating system does the work instead of the app.

What's your opinion on these two accessibility features?

To clarify the available options I've attached screenshots below.

# 1. Screen zoom category
## Android "Screen zoom"
<img alt="Screen zoom on Android at the lowest setting, 100%" src="https://github.com/user-attachments/assets/eff381c7-2bd9-40cf-a9f0-3aec01616526" width="200">
<img alt="Screen zoom on Android at the highest setting, close to 200%" src="https://github.com/user-attachments/assets/e2d5aec6-6e51-496d-8c08-b38bcbc08977" width="200">

## iOS "Display Zoom"
<img alt="Changing Display Zoom on iOS to the Larger Text setting" src="https://github.com/user-attachments/assets/a84f002f-9ab9-49d9-a57e-a5a778f4fdd8" width="200">
<img alt="Display Zoom on iOS at the Larger Text setting, around 150%" src="https://github.com/user-attachments/assets/1fdb19c5-7db8-43ec-98d7-34d4bbf3305d" width="200">

# 2. Screen magnify category
## Android "Magnification"
<img alt="Enabling Magnification on Android" src="https://github.com/user-attachments/assets/ecff0509-de6c-44e2-ab6b-3062bfa64d63" width="200">
<img alt="Zooming in a part of the screen on Android to 300%" src="https://github.com/user-attachments/assets/4203d3f2-4fbc-4a6a-b0fd-8699aaf891bc" width="200">

## iOS "Zoom"
<img alt="Zooming in like a magnifying glass on iOS" src="https://github.com/user-attachments/assets/9898f3fb-2cb3-4daa-adfd-e82584034c2a" width="200">
<img alt="Zooming on the whole screen on iOS to 400%" src="https://github.com/user-attachments/assets/38faa460-475a-420d-a530-171ea1ff3380" width="200">

**Comment by JJdeGroot** [comment]
In Chapter 11.7 of EN 301 549 v3.2.1 there are additional requirements, that also related to Resize Text

> Where software is not designed to be isolated from its platform, and provides a user interface, **that user interface shall
follow the values of the user preferences for platform settings for**: units of measurement, colour, contrast, font type, **font
size**, and focus cursor except where they are overridden by the user.

Meaning that in the normative text of WCAG2ICT, system zoom would be a way to pass 1.4.4, even if text does not scale.
But following the normative of EN 301 549, you would not pass because you would not honor the user preferences for font size.

cc @detlevhfischer - seems like no change for EN 301 549 V4.1.1b is not needed

**Comment by dotjay** [comment]
Does this warrant a new key term "platform settings" to incorporate such "user preferences", which would be an example of ["platform software"](https://w3c.github.io/wcag2ict/#platform-software)? It might also be a good thing to align with EN 301 549 terminology.

**Comment by JJdeGroot** [comment]
Related: https://github.com/w3c/wcag/issues/4290#issuecomment-2761758769
